File: examples/audio/audio.cpp
#include <math.h>
#include <array>

#include "picosystem.hpp"


using namespace picosystem;

// a place to store the current state of the dials
struct dial_t {
  std::string name, unit;
  int32_t min, max, value, step;
};

std::array<dial_t, 12> dials = {{
  {"frequency", "hz",    0, 8000, 440, 50},
  {   "volume",  "%",    0,  100, 100, 10},
  {  "sustain",  "%",    0,  100,  80, 10},
  {  "distort",  "%",    0,  100,   0, 10},
  {   "attack", "ms",    0, 1000, 100, 50},
  {    "decay", "ms",    0, 1000,  50, 50},
  { "duration", "ms",    0, 1000, 500, 50},
  {  "release", "ms",    0, 1000, 250, 50},
  {   "reverb", "ms",    0, 2000,   0, 50},
  {     "bend", "hz", -100,  100,   0, 10},
  {  "bend ms", "ms",    0, 1000, 100, 50},
  {    "noise",  "%",    0,  100,   0, 10}
}};

// index of the selected diable
uint32_t selected = 0;

// return a dial by name
dial_t *get_dial(std::string name) {
  for(auto &d :dials) {
    if(d.name == name) return &d;
  }
  return nullptr;
}

// change a dial value up or down and clamp to limits
void adjust_dial_value(dial_t &d, int32_t delta) {
  d.value += delta;
  d.value = d.value > d.max ? d.max : d.value;
  d.value = d.value < d.min ? d.min : d.value;
}

// return the current value of the named dial
uint32_t get_dial_value(std::string name) {
  dial_t *d = get_dial(name);
  return d->value;
}

// voice for the current dial settings
voice_t v;

// update and play the voice
void set_voice() {
  v = voice(
    get_dial_value(   "attack"),
    get_dial_value(    "decay"),
    get_dial_value(  "sustain"),
    get_dial_value(  "release"),
    get_dial_value(     "bend"),
    get_dial_value(  "bend ms"),
    get_dial_value(   "reverb"),
    get_dial_value(    "noise"),
    get_dial_value(  "distort"));

  play(v, get_dial_value("frequency"), get_dial_value("duration"), get_dial_value("volume"));
}

// initialise the world
void init() {
  set_voice();
}

// process user input and update the world state
void update(uint32_t tick) {
  bool change = false;

  // only check for UP/DOWN state every 10 ticks (100ms) to ensure that dial
  // values change at a sensible rate
  if(tick % 10 == 0) {
    if(button(UP)) {
      adjust_dial_value(dials[selected], dials[selected].step);
      change = true;
    }
    if(button(DOWN)) {
      adjust_dial_value(dials[selected], -dials[selected].step);
      change = true;
    }
  }

  // if LEFT or RIGHT has been pressed then move selection to the next or
  // previous dial
  if(pressed(LEFT)) {
    selected = selected == 0 ? dials.size() - 1 : selected - 1;
  }
  if(pressed(RIGHT)) {
    selected = selected < dials.size() - 1 ? selected + 1 : 0;
  }

  // if the current sound has finished playing, or we changed a dial value
  // then restart the sound
  uint32_t full_duration = get_dial_value("duration") + get_dial_value("release") + get_dial_value("reverb");
  if(change || position() > full_duration) {
    set_voice();
  }
}

void draw_dial(std::string name, int32_t x, int32_t y) {
  dial_t *d = get_dial(name);

  // draw outline
  pen(0, 0, 0, 2);
  rect(x, y, 60, 60);

  // draw rivets
  pen(0, 0, 0, 2);
  fcircle(x + 3, y + 3, 1);
  fcircle(x + 60 - 4, y + 3, 1);
  fcircle(x + 3, y + 60 - 4, 1);
  fcircle(x + 60 - 4, y + 60 - 3, 1);

  // draw dial face
  pen(0, 0, 0, 4);
  fcircle(x + 30, y + 20, 16);
  pen(15, 15, 15, 12);
  fcircle(x + 30, y + 20, 13);

  // draw dot of the dial
  float fv = float(d->value + abs(d->min)) / (d->max - d->min);
  int32_t dotx = -sin(3.1415927 * 0.2f + fv * 3.1415927 * 1.6f) * 9;
  int32_t doty = cos(3.1415927 * 0.2f + fv * 3.1415927 * 1.6f) * 9;
  pen(0, 0, 0, 4);
  fcircle(dotx + x + 30, doty + y + 20, 3);

  // centre dial label and value
  uint32_t lw = 60; //text_width(d->name);
  text(d->name, x + 30 - (lw / 2), y + 60 - 22);
  std::string unit_label = str(d->value) + d->unit;
  lw = 20;//text_width(unit_label);
  text(unit_label, x + 30 - (lw / 2), y + 60 - 12);

  // if this is the active dial then highlight it
  if(d == &dials[selected]) {
    pen(0, 15, 0);
    rect(x, y, 60, 60);
    rect(x + 1, y + 1, 58, 58);
  }
}

// draw the world
void draw(uint32_t tick) {
  // draw top row of dials
  pen(10, 12, 4);
  frect(0, 60, 240, 60);
  draw_dial("frequency", 0, 60);
  draw_dial("volume", 60, 60);
  draw_dial("sustain", 120, 60);
  draw_dial("distort", 180, 60);

  // draw middle row of dials
  pen(4, 10, 12);
  frect(0, 120, 240, 60);
  draw_dial("attack", 0, 120);
  draw_dial("decay", 60, 120);
  draw_dial("duration", 120, 120);
  draw_dial("release", 180, 120);

  // draw bottom row of dials
  pen(12, 4, 10);
  frect(0, 180, 240, 60);
  draw_dial("reverb", 0, 180);
  draw_dial("bend", 60, 180);
  draw_dial("bend ms", 120, 180);
  draw_dial("noise", 180, 180);

  // draw waveform graph
  pen(0, 0, 0, 2);
  frect(0, 0, 240, 60);
  uint32_t full_duration = get_dial_value("duration") + get_dial_value("release") + get_dial_value("reverb");
  pen(15, 15, 15);
  for(int i = 0; i < 230; i+=6) {
    uint8_t s = audio_sample((i * full_duration) / 230);
    frect(i + 2, 52 - s / 2, 5, 5);
  }

  // draw current playback marker
  pen(0, 15, 15);
  uint32_t pos = (position() * 230) / full_duration;
  frect(pos + 5, 5, 2, 50);
}

File: micropython/examples/picosystem/audio.py
import math


# a place to store the current state of the dials
class Dial:
    def __init__(self, name, unit, min, max, value, step, x, y):
        self.name = name
        self.unit = unit
        self.min = min
        self.max = max
        self.value = value
        self.step = step
        self.x = x
        self.y = y

    def adjust_value(self, v):
        self.value += v
        self.value = min(self.max, self.value)
        self.value = max(self.min, self.value)


dials = [
    Dial("frequency", "hz",    0, 8000, 440, 50, 0, 30),   # noqa E241
    Dial(   "volume",  "%",    0,  100, 100, 10, 30, 30),  # noqa E241
    Dial(  "sustain",  "%",    0,  100,  80, 10, 60, 30),  # noqa E241
    Dial(  "distort",  "%",    0,  100,   0, 10, 90, 30),  # noqa E241
    Dial(   "attack", "ms",    0, 1000, 100, 50, 0, 60),   # noqa E241
    Dial(    "decay", "ms",    0, 1000,  50, 50, 30, 60),  # noqa E241
    Dial( "duration", "ms",    0, 1000, 500, 50, 60, 60),  # noqa E241
    Dial(  "release", "ms",    0, 1000, 250, 50, 90, 60),  # noqa E241
    Dial(   "reverb", "ms",    0, 2000,   0, 50, 0, 90),   # noqa E241
    Dial(     "bend", "hz", -100,  100,   0, 10, 30, 90),  # noqa E241
    Dial(  "bend ms", "ms",    0, 1000, 100, 50, 60, 90),  # noqa E241
    Dial(    "noise",  "%",    0,  100,   0, 10, 90, 90)   # noqa E241
]

# index of the selected diable
selected = 0

# voice for the current dial settings
v = Voice()


def audio_position():
    return 0


def audio_sample(pos):
    return 0


# return a dial by name
def get_dial(name):
    for dial in dials:
        if dial.name == name:
            return dial
    return None


# change a dial value up or down and clamp to limits
def adjust_dial_value(d, delta):
    d.adjust_value(delta)


# return the current value of the named dial
def get_dial_value(name):
    dial = get_dial(name)
    if dial:
        return dial.value
    return 0


# update and play the voice
def set_voice():
    v.envelope(
        get_dial_value("attack"),
        get_dial_value("decay"),
        get_dial_value("sustain"),
        get_dial_value("release"),
    )
    v.bend(
        get_dial_value("bend"),
        get_dial_value("bend ms")
    )
    v.effects(
        get_dial_value("reverb"),
        get_dial_value("noise"),
        get_dial_value("distort")
    )
    v.play(get_dial_value("frequency"), get_dial_value("duration"), get_dial_value("volume"))


# process user input and update the world state
def update(tick):
    global selected
    change = False

    # only check for UP/DOWN state every 10 ticks (100ms) to ensure that dial
    # values change at a sensible rate
    if tick % 10 == 0:
        if button(UP):
            adjust_dial_value(dials[selected], dials[selected].step)
            change = True

        if button(DOWN):
            adjust_dial_value(dials[selected], -dials[selected].step)
            change = True

    # if LEFT or RIGHT has been pressed then move selection to the next or
    # previous dial
    if pressed(LEFT):
        selected -= 1

    if pressed(RIGHT):
        selected += 1

    selected %= len(dials)

    # if the current sound has finished playing, or we changed a dial value
    # then restart the sound
    full_duration = get_dial_value("duration") + get_dial_value("release") + get_dial_value("reverb")
    if change or audio_position() > full_duration:
        set_voice()


def draw_dial(name):
    d = get_dial(name)
    if d is None:
        return

    x = d.x
    y = d.y

    # draw outline
    pen(0, 0, 0, 2)
    rect(x, y, 30, 30)

    # draw rivets
    pen(0, 0, 0, 2)
    fcircle(x + 3, y + 3, 1)
    fcircle(x + 30 - 4, y + 3, 1)
    fcircle(x + 3, y + 30 - 4, 1)
    fcircle(x + 30 - 4, y + 30 - 3, 1)

    # draw dial face
    pen(0, 0, 0, 4)
    fcircle(x + 15, y + 15, 10)
    pen(15, 15, 15, 12)
    fcircle(x + 15, y + 15, 8)

    # draw dot of the dial
    fv = float(d.value + abs(d.min)) / (d.max - d.min)
    dotx = -int(math.sin(3.1415927 * 0.2 + fv * 3.1415927 * 1.6) * 5)
    doty = int(math.cos(3.1415927 * 0.2 + fv * 3.1415927 * 1.6) * 5)
    pen(0, 0, 0, 4)
    fcircle(dotx + x + 15, doty + y + 15, 2)

    # if this is the active dial then highlight it
    if d == dials[selected]:
        pen(0, 15, 0)
        rect(x, y, 30, 30)
        rect(x + 1, y + 1, 28, 28)


# draw the world
def draw(tick):
    # draw top row of dials
    pen(10, 12, 4)
    frect(0, 30, 120, 30)
    draw_dial("frequency")
    draw_dial("volume")
    draw_dial("sustain")
    draw_dial("distort")

    # draw middle row of dials
    pen(4, 10, 12)
    frect(0, 60, 120, 30)
    draw_dial("attack")
    draw_dial("decay")
    draw_dial("duration")
    draw_dial("release")

    # draw bottom row of dials
    pen(12, 4, 10)
    frect(0, 90, 120, 30)
    draw_dial("reverb")
    draw_dial("bend")
    draw_dial("bend ms")
    draw_dial("noise")

    d = dials[selected]
    # show dial label and value
    x = d.x
    if x >= 60:
        x -= 90
    pen(15, 15, 15, 13)
    frect(x + 30, d.y + 30 - 20, 60, 20)
    pen(0, 0, 0, 4)
    text(d.name, x + 30 + 2, d.y + 30 - 18)
    unit_label = f"{d.value}{d.unit}"
    text(unit_label, x + 30 + 2, d.y + 30 - 10)

    # draw waveform graph
    pen(0, 0, 0, 2)
    frect(0, 0, 120, 30)
    full_duration = get_dial_value("duration") + get_dial_value("release") + get_dial_value("reverb")
    pen(15, 15, 15)
    for i in range(0, 115, 6):
        s = audio_sample(int((i * full_duration) / 115))
        frect(i + 2, 25 - int(s / 2), 5, 5)

    # draw current playback marker
    pen(0, 15, 15)
    pos = int((audio_position() * 115) / full_duration)
    frect(pos + 5, 5, 2, 25)


set_voice()
start()
